### ISSUE TITLE
[Bug]: Scalar indexes cannot search out data

### ISSUE THREAD

**Issue by syang1997** [milvus-io/milvus#34548]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: v2.3.15
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka):   kafka 
- SDK version(e.g. pymilvus v2.0.0rc2): java sdk 2.3.4  , attu
- OS(Ubuntu or CentOS): CentOS
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

Hybrid search cannot find out data, but a separate query can find out data
![img_3](https://github.com/milvus-io/milvus/assets/48277927/f7bc02b7-61de-4564-b176-c9b753f766c0)
![img_4](https://github.com/milvus-io/milvus/assets/48277927/95e5651e-1560-45bf-a264-f3ea0764b97e)
This scalar query has data
![img_5](https://github.com/milvus-io/milvus/assets/48277927/bc4cfbc5-d69e-43ed-b972-63c379cfc1a2)

### Milvus Log

[milvus-log (1).tar.gz](https://github.com/user-attachments/files/16144447/milvus-log.1.tar.gz)

**Comment by syang1997** [comment]
Just now I backed up this vector to another cluster, and the exception can still be reproduced. After deleting the collection from the backup cluster, I backed up and restored it again, and it cannot be reproduced.

**Comment by yanliang567** [comment]
/assign @zhagnlu 
please help to take a look
/unassign

**Comment by syang1997** [comment]
This is not because scalar filtering causes hnsw to be unable to perform layer traversal, because there is another set of data that can also have data with scalars after multiple queries. And no data operation is performed during this period

**Comment by syang1997** [comment]
![企业微信截图_0d95f1b9-6c93-432b-a642-65ab25f5e055](https://github.com/milvus-io/milvus/assets/48277927/adc24d35-fc1d-4db2-b1c5-2d162989c4d3)
![企业微信截图_62c52532-acb0-4f40-8fca-acf446495788](https://github.com/milvus-io/milvus/assets/48277927/237b894f-6697-48a2-9524-ac1566ada3e1)

**Comment by syang1997** [comment]
This collection configures replicas. Is it caused by index differences between replicas?

**Comment by syang1997** [comment]
> what difference between upper and bottom search ?
There is no difference, multiple requests return completely different results

**Comment by syang1997** [comment]
> > > what difference between upper and bottom search ?
> > 
> > 
> > @zhagnlu There is no difference, multiple requests return completely different results
> 
> if not hybrid search, just using query, will multiple requests return completely different results ?

Vector search returns normally

**Comment by syang1997** [comment]
@zhagnlu Another phenomenon is that some search conditions cannot be returned at all if they have scalar filtering, but vector searches have returns. But this scalar filtering has data. hybrid search returns blank

**Comment by yanliang567** [comment]
okay, so the issue here is that if using query with expr filter on scalar fields, the results are not correct or consistent(not expected). But if query without fitlering, the results are always consistent(expected). Am I right?
/assign @cydrain @liliu-z 
could you please also help to take a look

**Comment by syang1997** [comment]
@yanliang567 Yes, sometimes the returned results are inconsistent, and sometimes the returned results are incorrect.

**Comment by syang1997** [comment]
Regenerated debug log
[milvus-log (2).tar.gz](https://github.com/user-attachments/files/16232353/milvus-log.2.tar.gz)

**Comment by alwayslove2013** [comment]
@syang Could you please tell us the filter_rate and index building parameters?
The current open source Milvus may have “less than top-k search results” problems with high filter_rates (70-90%) and low `M`.

**Comment by syang1997** [comment]
> @syang Could you please tell us the filter_rate and index building parameters? The current open source Milvus may have “less than top-k search results” problems with high filter_rates (70-90%) and low `M`.

<img width="1502" alt="image" src="https://github.com/user-attachments/assets/1eeb8fb6-2236-46f4-8ab2-aeb2eb82a408">
Most searches that have been adjusted before are normal, but now the M value is not small

**Comment by syang1997** [comment]
@alwayslove2013 This collection has less than 20,000 data, but the M value and efConstruction are large enough (I think).I know about the data island problem that scalar filtering and hnsw work together, and I have previously investigated and adjusted the index construction parameters

**Comment by cydrain** [comment]
Hi @syang1997 ,

Can you share your script to reproduce this issue ?

**Comment by syang1997** [comment]
> Hi @syang1997 ,
> 
> Can you share your script to reproduce this issue ?

I'm coding a demo to replicate this issue

**Comment by cydrain** [comment]
Hi @syang1997 ,

One more question, I see you're using Milvus v2.3.15, have you tried Milvus v2.4.x ?

**Comment by xiaofan-luan** [comment]
@syang1997 
I think we need data to reproduce this issue.
@cydrain please setup a meeting with syang see if we can get some data to reproduce

**Comment by syang1997** [comment]
> @syang1997 I think we need data to reproduce this issue. @cydrain please setup a meeting with syang see if we can get some data to reproduce

We have already communicated with the community once, and the preliminary reason is still the previous data island problem

**Comment by syang1997** [comment]
> @syang1997 I think we need data to reproduce this issue. @cydrain please setup a meeting with syang see if we can get some data to reproduce

@xiaofan-luan The phenomenon is that there is no return instead of returning insufficient topK, so it is suspected that the first layer node of HNSW is filtered by all

**Comment by xiaofan-luan** [comment]
after discussion, it seems the reason might be hnsw filtered 70-80% data， cause graph connectivity brokes

**Comment by liliu-z** [comment]
This fix will be released with 2.4.7

**Comment by syang1997** [comment]
> This fix will be released with 2.4.7

Can it be merged to version 2.3.x?

**Comment by yanliang567** [comment]
> > This fix will be released with 2.4.7
> 
> Can it be merged to version 2.3.x? @yanliang567 @liliu-z

I don't think so, as 2.3.20 could be the last release of 2.3.x

**Comment by syang1997** [comment]
> > > This fix will be released with 2.4.7
> > 
> > 
> > Can it be merged to version 2.3.x? @yanliang567 @liliu-z
> 
> I don't think so, as 2.3.20 could be the last release of 2.3.x

Okay, then we will choose to upgrade to 2.4.x later
